Add terminal-based steering interface via an extract

siramok opened the pull request:
This PR adds a new extract which puts the user into a basic shell-like steering interface. The main benefit is to have a simpler, no-dependency alternative to the existing ascent-jupyter-bridge steering interface.

It can be accessed via:
```-
  action: "add_extracts"
  extracts:
    e1:
      type: "steering"
```
The current implementation allows users to manually, at runtime:
- Add, edit, or delete numeric/string parameters with which to run callbacks (via an internally managed conduit node).
- List registered callbacks (via new APIs that expose the callback names).
- Run callbacks by name.
- See any resulting output from running a callback.

Still to-do:
- [x] Verify that all MPI usage is properly guarded by ASCENT_MPI_ENABLED.
- [x] Similarly, verify that things work properly without MPI.
- [x] Add tests.

Despite not being ready, the big picture implementation details are unlikely to change (where I've implemented it, how I've plumbed it through). Any feedback you may have for me is welcome.

siramok committed:
Put some foundation in place

siramok committed:
Cleanup output messages

siramok committed:
Forgot to guard the mpi.h include

cyrush commented on src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp:
```diff
@@ -0,0 +1,389 @@
+    descriptions["run"] = "Run an Ascent callback with 'run my_callback_name'.";
+
+#ifdef ASCENT_MPI_ENABLED
+    MPI_Comm_rank(MPI_COMM_WORLD, &m_rank);
```
We might not be using COMM_WORLD, so let's grab the right communicator using:
```
MPI_Comm mpi_comm = MPI_Comm_f2c(flow::Workspace::default_mpi_comm());
```

cyrush commented on src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.hpp:
```diff
@@ -0,0 +1,94 @@
+    std::map<std::string, std::function<void()>> commands;
+    std::map<std::string, std::string> descriptions;
+    conduit::Node params;
+    conduit::Node output;
```
We like to use `m_` prefix for member vars

cyrush commented on src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.hpp:
```diff
@@ -0,0 +1,94 @@
+
+    void empty_run();
+    void exit_shell();
+    void parse_input(std::string cmd, std::vector<std::string> args);
```
```suggestion
    void parse_input(std::string &cmd, std::vector<std::string> &args);
```

cyrush commented on src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.hpp:
```diff
@@ -0,0 +1,94 @@
+    void exit_shell();
+    void parse_input(std::string cmd, std::vector<std::string> args);
+    void list_callbacks();
+    void modify_params(std::vector<std::string> args);
```
```suggestion
    void modify_params(std::vector<std::string> &args);
```

cyrush commented on src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.hpp:
```diff
@@ -0,0 +1,94 @@
+    void modify_params(std::vector<std::string> args);
+    void print_help();
+    void print_params();
+    void run_callback(std::vector<std::string> args);
```
```suggestion
    void run_callback(std::vector<std::string> &args);
```

cyrush commented on src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp:
```diff
@@ -0,0 +1,389 @@
+
+        // This might not be necessary?
+#ifdef ASCENT_MPI_ENABLED
+        MPI_Barrier(MPI_COMM_WORLD);
```
fetch comp from workspace

cyrush commented on src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp:
```diff
@@ -0,0 +1,389 @@
+    }
+}
+
+void Steering::run_callback(std::vector<std::string> args)
```
```suggestion
void Steering::run_callback(std::vector<std::string> &args)
```

cyrush commented on src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp:
```diff
@@ -0,0 +1,389 @@
+    }
+}
+
+void Steering::modify_params(std::vector<std::string> args)
```
```suggestion
void Steering::modify_params(std::vector<std::string> &args)
```

cyrush commented on src/libs/ascent/runtimes/flow_filters/ascent_runtime_steering_filters.cpp:
```diff
@@ -0,0 +1,389 @@
+        }
+
+#ifdef ASCENT_MPI_ENABLED
+        MPI_Bcast(&input_size, 1, MPI_INT, 0, MPI_COMM_WORLD);
```
grab comm from workspace

cyrush commented on src/libs/ascent/ascent.hpp:
```diff
@@ -87,6 +87,14 @@ void        ASCENT_API execute_callback(std::string callback_name,
 bool        ASCENT_API execute_callback(std::string callback_name);
 
+//-----------------------------------------------------------------------------
+std::vector<std::string>
```
ref copy out style will be slight more efficient:
```
void ASCENT_API get_void_callbacks(std::vector<std::string> &name);
```

cyrush commented on src/libs/ascent/ascent.hpp:
```diff
@@ -87,6 +87,14 @@ void        ASCENT_API execute_callback(std::string callback_name,
+
+//-----------------------------------------------------------------------------
+std::vector<std::string>
+            ASCENT_API get_bool_callbacks();
```
ref copy out style will be slight more efficient:
```
void ASCENT_API get_bool_callbacks(std::vector<std::string> &name);
```

cyrush commented on src/libs/ascent/ascent.hpp:
```diff
@@ -87,6 +87,14 @@ void        ASCENT_API execute_callback(std::string callback_name,
 //-----------------------------------------------------------------------------
 bool        ASCENT_API execute_callback(std::string callback_name);
```
```suggestion
bool        ASCENT_API execute_callback(const std::string &callback_name);
```

cyrush commented on src/libs/ascent/ascent.cpp:
```diff
@@ -928,6 +928,28 @@ execute_callback(std::string callback_name)
 }
 
+//-----------------------------------------------------------------------------
+std::vector<std::string>
```
copy out still ref args are a bit more efficient (see below)

cyrush reviewed (commented):
Look great!
a few minor suggestions related to using C++ refs for strings and vectors, as well as the recipe for grabbing the right MPI comm

siramok committed:
Address feedback

cyrush reviewed (approved):
thanks for the tweaks!

siramok committed:
Add tests, formatting pass

cyrush commented:
Tests exercising steering look good -  let me know when you feel its ready to merge.

siramok committed:
Polishing, improvements

siramok commented:
Sweet, thanks Cyrus. I just made another polishing pass. Added comments, fixed formatting, simplified logic where I could, and made some tweaks after testing it out for myself. I feel comfortable merging it in it's current state.

I'll go create a new branch to work on an example instrumentation which makes use of this.

siramok committed:
Merge branch 'Alpine-DAV:develop' into task/terminal-steering
